File: process_12_categories.py
#!/usr/bin/env python3
"""
Translate JSON files to bilingual format for 12 specific categories
"""

import json
import re
import time
from pathlib import Path
from openai import OpenAI

# Initialize OpenAI client
client = OpenAI()

def detect_language(text):
    """Detect if text is primarily Chinese or English"""
    if not text:
        return 'en'
    chinese_chars = len(re.findall(r'[一-鿿]', text))
    english_chars = len(re.findall(r'[a-zA-Z]', text))
    return 'zh' if chinese_chars > english_chars * 0.5 else 'en'

def translate_text(text, field_type='content'):
    """Translate text to target language using OpenAI"""
    if not text:
        return text

    # Check if already in target language
    lang = detect_language(text)

    # Truncate for API limits
    max_tokens = 4000 if field_type == 'content' else 500
    text_to_translate = text[:max_tokens * 4]

    try:
        if lang == 'zh':
            # Translate Chinese to English
            prompt = f"Translate this Chinese {field_type} to English. Maintain all HTML structure and formatting. Output only the translation:"
        else:
            # Translate English to Chinese
            prompt = f"Translate this English {field_type} to Chinese. Maintain all HTML structure and formatting. Output only the translation:"

        response = client.chat.completions.create(
            model="gpt-4o-mini",
            messages=[
                {"role": "system", "content": prompt},
                {"role": "user", "content": text_to_translate}
            ],
            max_tokens=max_tokens,
            temperature=0.3
        )
        return response.choices[0].message.content.strip()
    except Exception as e:
        print(f"Translation error for {field_type}: {e}")
        return text

def translate_keywords(keywords, source_lang):
    """Translate keywords array"""
    if not keywords or not isinstance(keywords, list):
        return keywords

    translated = []
    for kw in keywords:
        if not kw:
            translated.append(kw)
            continue

        kw_lang = detect_language(kw)

        # Skip if already in target language
        if source_lang == 'zh' and kw_lang == 'en':
            translated.append(kw)
            continue
        elif source_lang == 'en' and kw_lang == 'zh':
            translated.append(kw)
            continue

        try:
            prompt = f"Translate this SEO keyword from {kw_lang} to {'en' if source_lang == 'zh' else 'zh'}. Output only the translated keyword:"
            response = client.chat.completions.create(
                model="gpt-4o-mini",
                messages=[
                    {"role": "system", "content": prompt},
                    {"role": "user", "content": kw}
                ],
                max_tokens=50,
                temperature=0.3
            )
            translated.append(response.choices[0].message.content.strip())
        except Exception:
            translated.append(kw)

    return translated

def process_file(file_path):
    """Process a single JSON file to bilingual format"""
    try:
        with open(file_path, 'r', encoding='utf-8') as f:
            data = json.load(f)

        # Skip if already bilingual
        if 'title_zh' in data or 'title_en' in data:
            return "already bilingual"

        # Detect original language
        original_lang = detect_language(data.get('title', '') or data.get('content', ''))

        if original_lang == 'zh':
            # Chinese file - add English translation
            data['title_zh'] = data.get('title', '')
            data['title_en'] = translate_text(data.get('title', ''), 'title')

            data['description_zh'] = data.get('description', '')
            data['description_en'] = translate_text(data.get('description', ''), 'description')

            data['content_zh'] = data.get('content', '')
            data['content_en'] = translate_text(data.get('content', ''), 'content')

            # Translate keywords
            keywords_zh = data.get('seo_keywords', [])
            data['seo_keywords_zh'] = keywords_zh
            data['seo_keywords_en'] = translate_keywords(keywords_zh, 'zh')

            # Remove old single-language fields
            data.pop('title', None)
            data.pop('description', None)
            data.pop('content', None)
            data.pop('seo_keywords', None)
        else:
            # English file - add Chinese translation
            data['title_en'] = data.get('title', '')
            data['title_zh'] = translate_text(data.get('title', ''), 'title')

            data['description_en'] = data.get('description', '')
            data['description_zh'] = translate_text(data.get('description', ''), 'description')

            data['content_en'] = data.get('content', '')
            data['content_zh'] = translate_text(data.get('content', ''), 'content')

            # Translate keywords
            keywords_en = data.get('seo_keywords', [])
            data['seo_keywords_en'] = keywords_en
            data['seo_keywords_zh'] = translate_keywords(keywords_en, 'en')

            # Remove old single-language fields
            data.pop('title', None)
            data.pop('description', None)
            data.pop('content', None)
            data.pop('seo_keywords', None)

        # Write back
        with open(file_path, 'w', encoding='utf-8') as f:
            json.dump(data, f, ensure_ascii=False, indent=2)

        return "success"

    except Exception as e:
        return f"error: {str(e)}"

def main():
    # Categories to process
    categories = [
        "bike-cycling-rental-tools",
        "blue-collar-tools",
        "boat-marine-rental-tools",
        "camera-photography-rental-tools",
        "camping-outdoor-gear-rental-tools",
        "car-vehicle-rental-tools",
        "casino-gaming-entertainment-tools",
        "child-care-preschool-tools",
        "cleaning-maintenance-rental-tools",
        "construction-building-rental-tools",
        "construction-contractor-tools",
        "costume-fashion-rental-tools"
    ]

    base_dir = Path("/Users/gejiayu/owner/seo/data")

    # Collect all files
    all_files = []
    for category in categories:
        cat_dir = base_dir / category
        if cat_dir.exists():
            for json_file in cat_dir.glob("*.json"):
                all_files.append(json_file)

    print(f"Total files to process: {len(all_files)}")
    print(f"Categories: {categories}")
    print()

    # Process files sequentially
    success_count = 0
    skip_count = 0
    error_count = 0

    for i, file_path in enumerate(all_files):
        status = process_file(file_path)

        if status == "success":
            success_count += 1
        elif status == "already bilingual":
            skip_count += 1
        else:
            error_count += 1
            print(f"❌ Error processing {file_path.name}: {status}")

        # Report every 20 files
        if (i + 1) % 20 == 0:
            print(f"✓ Progress: {i+1}/{len(all_files)} | Success: {success_count} | Skip: {skip_count} | Error: {error_count}")

        # Rate limit delay
        time.sleep(0.5)

    print()
    print("=" * 80)
    print("FINAL RESULTS")
    print("=" * 80)
    print(f"Total processed: {len(all_files)}")
    print(f"Success: {success_count}")
    print(f"Skipped (already bilingual): {skip_count}")
    print(f"Errors: {error_count}")
    print("=" * 80)

if __name__ == "__main__":
    main()

File: process_12_categories.mjs
#!/usr/bin/env node
/**
 * Translate JSON files in 12 specified categories to bilingual format
 * Uses OpenAI GPT-4o-mini for accurate translation
 */

import fs from 'fs';
import path from 'path';
import OpenAI from 'openai';
import { fileURLToPath } from 'url';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

// Initialize OpenAI
const openai = new OpenAI();

// Categories to process
const categories = [
  'bike-cycling-rental-tools',
  'blue-collar-tools',
  'boat-marine-rental-tools',
  'camera-photography-rental-tools',
  'camping-outdoor-gear-rental-tools',
  'car-vehicle-rental-tools',
  'casino-gaming-entertainment-tools',
  'child-care-preschool-tools',
  'cleaning-maintenance-rental-tools',
  'construction-building-rental-tools',
  'construction-contractor-tools',
  'costume-fashion-rental-tools'
];

const dataDir = '/Users/gejiayu/owner/seo/data';

/**
 * Detect if text is primarily Chinese
 */
function isChinese(text) {
  if (!text) return false;
  const chineseChars = (text.match(/[一-鿿]/g) || []).length;
  const englishChars = (text.match(/[a-zA-Z]/g) || []).length;
  return chineseChars > englishChars * 0.5;
}

/**
 * Translate text using OpenAI
 */
async function translateText(text, fieldType = 'content') {
  if (!text) return text;

  // Skip if already in target language
  const lang = isChinese(text) ? 'zh' : 'en';

  // Truncate for API limits
  const maxTokens = fieldType === 'content' ? 4000 : 500;
  const textToTranslate = text.substring(0, maxTokens * 4);

  try {
    const targetLang = lang === 'zh' ? 'English' : 'Chinese';
    const prompt = `Translate this ${fieldType} from ${lang === 'zh' ? 'Chinese' : 'English'} to ${targetLang}. Maintain all HTML structure and formatting exactly. Output only the translation:`;

    const response = await openai.chat.completions.create({
      model: 'gpt-4o-mini',
      messages: [
        { role: 'system', content: prompt },
        { role: 'user', content: textToTranslate }
      ],
      max_tokens: maxTokens,
      temperature: 0.3
    });

    return response.choices[0].message.content.trim();
  } catch (error) {
    console.error(`Translation error for ${fieldType}: ${error.message}`);
    return text;
  }
}

/**
 * Translate keywords array
 */
async function translateKeywords(keywords, sourceLang) {
  if (!keywords || !Array.isArray(keywords)) return keywords;

  const translated = [];
  for (const kw of keywords) {
    if (!kw) {
      translated.push(kw);
      continue;
    }

    const kwLang = isChinese(kw) ? 'zh' : 'en';

    // Skip if already in target language
    if (sourceLang === 'zh' && kwLang === 'en') {
      translated.push(kw);
      continue;
    } else if (sourceLang === 'en' && kwLang === 'zh') {
      translated.push(kw);
      continue;
    }

    try {
      const targetLang = kwLang === 'zh' ? 'English' : 'Chinese';
      const response = await openai.chat.completions.create({
        model: 'gpt-4o-mini',
        messages: [
          {
            role: 'system',
            content: `Translate this SEO keyword from ${kwLang === 'zh' ? 'Chinese' : 'English'} to ${targetLang}. Output only the translated keyword:`
          },
          { role: 'user', content: kw }
        ],
        max_tokens: 50,
        temperature: 0.3
      });

      translated.push(response.choices[0].message.content.trim());
    } catch (error) {
      translated.push(kw);
    }

    // Small delay between keywords
    await sleep(0.1);
  }

  return translated;
}

/**
 * Sleep helper
 */
function sleep(seconds) {
  return new Promise(resolve => setTimeout(resolve, seconds * 1000));
}

/**
 * Process single file to bilingual format
 */
async function processFile(filePath) {
  try {
    const data = JSON.parse(fs.readFileSync(filePath, 'utf-8'));

    // Skip if already bilingual
    if (data.title_zh || data.title_en) {
      return 'already bilingual';
    }

    // Detect original language
    const originalLang = isChinese(data.title || data.content || '') ? 'zh' : 'en';

    if (originalLang === 'zh') {
      // Chinese file - add English translation
      data.title_zh = data.title || '';
      data.title_en = await translateText(data.title, 'title');

      data.description_zh = data.description || '';
      data.description_en = await translateText(data.description, 'description');

      data.content_zh = data.content || '';
      data.content_en = await translateText(data.content, 'content');

      // Translate keywords
      const keywordsZh = data.seo_keywords || [];
      data.seo_keywords_zh = keywordsZh;
      data.seo_keywords_en = await translateKeywords(keywordsZh, 'zh');

      // Remove old fields
      delete data.title;
      delete data.description;
      delete data.content;
      delete data.seo_keywords;
    } else {
      // English file - add Chinese translation
      data.title_en = data.title || '';
      data.title_zh = await translateText(data.title, 'title');

      data.description_en = data.description || '';
      data.description_zh = await translateText(data.description, 'description');

      data.content_en = data.content || '';
      data.content_zh = await translateText(data.content, 'content');

      // Translate keywords
      const keywordsEn = data.seo_keywords || [];
      data.seo_keywords_en = keywordsEn;
      data.seo_keywords_zh = await translateKeywords(keywordsEn, 'en');

      // Remove old fields
      delete data.title;
      delete data.description;
      delete data.content;
      delete data.seo_keywords;
    }

    // Write back
    fs.writeFileSync(filePath, JSON.stringify(data, null, 2), 'utf-8');

    return 'success';
  } catch (error) {
    return `error: ${error.message}`;
  }
}

/**
 * Main execution
 */
async function main() {
  console.log('========================================');
  console.log('Bilingual Translation for 12 Categories');
  console.log('========================================\n');

  // Collect all files
  const allFiles = [];
  for (const category of categories) {
    const catDir = path.join(dataDir, category);
    if (fs.existsSync(catDir)) {
      const files = fs.readdirSync(catDir)
        .filter(f => f.endsWith('.json'))
        .map(f => path.join(catDir, f));
      allFiles.push(...files);
    }
  }

  console.log(`Total files to process: ${allFiles.length}`);
  console.log(`Categories: ${categories.join(', ')}`);
  console.log('\n');

  // Process files
  let successCount = 0;
  let skipCount = 0;
  let errorCount = 0;

  for (let i = 0; i < allFiles.length; i++) {
    const filePath = allFiles[i];
    const fileName = path.basename(filePath);

    const status = await processFile(filePath);

    if (status === 'success') {
      successCount++;
    } else if (status === 'already bilingual') {
      skipCount++;
    } else {
      errorCount++;
      console.log(`❌ Error: ${fileName} - ${status}`);
    }

    // Report every 20 files
    if ((i + 1) % 20 === 0) {
      console.log(`✓ Progress: ${i + 1}/${allFiles.length} | Success: ${successCount} | Skip: ${skipCount} | Error: ${errorCount}`);
    }

    // Rate limit delay
    await sleep(0.5);
  }

  console.log('\n========================================');
  console.log('FINAL RESULTS');
  console.log('========================================');
  console.log(`Total processed: ${allFiles.length}`);
  console.log(`Success: ${successCount}`);
  console.log(`Skipped: ${skipCount}`);
  console.log(`Errors: ${errorCount}`);
  console.log('========================================\n');
}

// Run
main().catch(console.error);